File: Emisor/fletcher-16.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <bitset>
#include <cstdlib>
#include <ctime>
#include <winsock2.h>

using namespace std;

uint16_t fletcher16(const vector<uint8_t>& data) {
    int sum1 = 0;
    int sum2 = 0;

    for (size_t i = 0; i < data.size(); ++i) {
        sum1 = (sum1 + data[i]) % 255;
        sum2 = (sum2 + sum1) % 255;
    }

    return (sum2 << 8) | sum1;
}

string checksumToBinary(uint16_t checksum) {
    string binary = "";
    for (int i = 15; i >= 0; --i) {
        binary += (checksum & (1 << i)) ? '1' : '0';
    }
    return binary;
}

// Función para convertir una cadena a binario
string stringToBinary(const string& input) {
    string binaryString;

    for (size_t i = 0; i < input.size(); ++i) {
        char c = input[i];
        bitset<8> binary(c);
        binaryString += binary.to_string();
    }
    
    return binaryString;
}

string fletcherCode() {
    string binaryMessage, originalMessage, codedMessage, noisyMessage;
    bool flag = false;

    // Inicializa la semilla del generador de números aleatorios
    srand(static_cast<unsigned int>(time(0)));

    cout << "--------------- Codigo Fletcher 16 ---------------" << endl << endl;

    while (!flag) {
        cout << "Ingrese la cadena a utilizar: " << endl;
        getline(cin, originalMessage);

        if (originalMessage.length() > 0) { //Si la cadena es mayor a 0
            binaryMessage = stringToBinary(originalMessage); //Convierte la cadena a binario

            // Eliminar los espacios en blanco al final de binaryMessage
            binaryMessage.erase(find_if(binaryMessage.rbegin(), binaryMessage.rend(), [](unsigned char ch) {
                return !isspace(ch);
            }).base(), binaryMessage.end());

            if (binaryMessage.length() % 8 != 0) { //Si la cadena no es múltiplo de 8
                while (binaryMessage.length() % 8 != 0) {
                    binaryMessage = '0' + binaryMessage;
                }
            }

            flag = true;
        } else {
            cout << "La longitud de la cadena binaria debe ser mayor a 0, intente nuevamente." << endl;
        }
    }

    cout << "Cadena a usar: " << binaryMessage << endl;

    vector<uint8_t> message;

    for (char c : binaryMessage) {
        message.push_back(c == '0' ? 0 : 1);
    }

    uint16_t checksum = fletcher16(message);
    string checksumBinary = checksumToBinary(checksum);

    codedMessage = binaryMessage + checksumBinary;
    
    cout << endl << "--------------- Codigo Fletcher-16 ---------------" << endl;

    for (char c : codedMessage) {
        int randomNumber = rand() % 100;

        if (randomNumber == 0) {
            if (c == '0') { //Cambia el valor
                c = '1';
            } else {
                c = '0';
            }

            cout << "\033[31m" << c << "\033[0m"; //Color rojo cambio de bit
        } else {
            cout << c; //Color negro bit normal
        }

        noisyMessage += c;
    }

    return noisyMessage;
}

int main() {
    // Initialize Winsock
    WSADATA wsaData;
    WSAStartup(MAKEWORD(2, 2), &wsaData);

    // Create a socket
    SOCKET s = socket(AF_INET, SOCK_STREAM, 0);
    if (s == INVALID_SOCKET) {
        cerr << "Could not create socket: " << WSAGetLastError() << endl;
        return 1;
    }

    // Setup the sockaddr_in structure
    sockaddr_in server;
    server.sin_family = AF_INET;
    server.sin_addr.s_addr = inet_addr("127.0.0.1");
    server.sin_port = htons(12345);

    // Connect to the server
    if (connect(s, (sockaddr*)&server, sizeof(server)) < 0) {
        cerr << "Connection failed: " << WSAGetLastError() << endl;
        return 1;
    }

    string noisyMessage = fletcherCode();
    
    cout<<noisyMessage.c_str();

    // Send data
    if (send(s, noisyMessage.c_str(), noisyMessage.size(), 0) < 0) {
        cerr << "Send failed: " << WSAGetLastError() << endl;
        return 1;
    }


    closesocket(s);
    WSACleanup();

    return 0;
}

File: Receptor/fletcher-16.py
def binaryToChecksum(binario):
    decimal = 0
    longitud = len(binario)

    for i in range(longitud):
        bit = int(binario[i])
        decimal += bit * (2 ** (longitud - 1 - i))

    return decimal

def fletcher16(data):
    sum1 = 0
    sum2 = 0

    for byte in data:
        sum1 = (sum1 + int(byte)) % 255
        sum2 = (sum2 + sum1) % 255

    return (sum2 << 8) | sum1

def binary_to_text(binary_str):
    # Divide el mensaje binario en bloques de 8 bits
    n = int(binary_str, 2)
    binary_list = [binary_str[i:i+8] for i in range(0, len(binary_str), 8)]
    # Convierte cada bloque de 8 bits a un carácter ASCII
    text = ''.join([chr(int(b, 2)) for b in binary_list])
    return text

def fletcherCode(binaryMessage):
    flag = False

    # print("\n\n--------------- Receptor en codigo Fletcher-16 ---------------\n")

    while not flag:
        # binaryMessage = input("Ingrese la cadena a utilizar: \n")

        if(len(binaryMessage)>0):
            if(len(binaryMessage) % 8 != 0): #Si la longitud de la cadena no es múltiplo de 8
                print("\nLa longitud del mensaje es de ", len(binaryMessage), " y debe ser un multiplo de 8, intente nuevamente.")
                print("---------------------------------------------------------------------------\n\n")

            else: #Si la longitud de la cadena es múltiplo de 8
                flag = True 
        else:
            print("\nLa longitud del mensaje debe ser mayor a 0, intente nuevamente.")
            print("---------------------------------------------------------------------------\n\n")

    binaryChecksum = binaryMessage[-16:]
    message = binaryMessage[:-16]

    checksumFromBinary = binaryToChecksum(binaryChecksum)
    checksumFromMessage = fletcher16(message)

    if(checksumFromBinary == checksumFromMessage): #Si los checksum's coinciden
        print("\n\n--------------- Mensaje original --------------- \n", {binary_to_text(message)})

    else:
        print("\n\n**************** Se detecto un error **************** \n")
        print("*************** El mensaje se descarta *************** \n")

def main():
    import socket

    # Configuration
    host = 'localhost'
    port = 12345

    # Create a TCP/IP socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.bind((host, port))
    sock.listen(1)

    print(f"Listening on {host}:{port}...")

    while True:
        connection, client_address = sock.accept()
        try:
            print(f"Connection from {client_address}")

            # Receive the data in small chunks
            data = b""
            while True:
                chunk = connection.recv(4096)
                if not chunk:
                    break
                data += chunk

            if data:
                binaryMessage = data.decode()
                fletcherCode(binaryMessage)
        finally:
            connection.close()
if __name__ == "__main__":
    main()
